evm/src/fixed_recursive_verifier.rs

```diff
@@ -787,13 +802,23 @@ where
         // Connect intermediary values for gas_used and bloom filters to the block's final values. We only plug on the right, so there is no need to check the left-handside block.
         Self::connect_final_block_values_to_intermediary(builder, rhs);
 
-        // Chack that the genesis block number is 0.
         let zero = builder.zero();
         let has_not_parent_block = builder.sub(one, has_parent_block.target);
+        // Chack that the genesis block number is 0.
         let gen_block_constr = builder.mul(has_not_parent_block, rhs.block_metadata.block_number);
         builder.connect(gen_block_constr, zero);
 
-        // TODO: Check that the genesis block has a predetermined state trie root.
+        // Check that the genesis block has a predetermined state trie root.
+        for (&limb0, limb1) in rhs
+            .trie_roots_before
+            .state_root
+            .iter()
+            .zip(rhs.extra_block_data.genesis_state_root)
+        {
+            let mut constr = builder.sub(limb0, limb1);
+            constr = builder.mul(has_not_parent_block, constr);
+            builder.connect(constr, zero);
+        }
     }
```

Comment:
Maybe this could be grouped into a dedicated genesis_block related function for clarity?